SUNDIALS updates

Steven-Roberts opened the pull request:
## Description
 
Replaces some deprecated SUNDIALS calls, adds a missing `N_VDestroy`, fixes uninitialized member variables that could cause issues when freed, and simplifies some of the code in the Ida and Kinsol classes. There's also some TODOs I left with questions @pelesh 

 ## Checklist
 
- [x] All tests pass.
- [x] Code compiles cleanly with flags `-Wall -Wpedantic -Wconversion -Wextra`.
- [x] The new code follows GridKit™ style guidelines.
- [ ] There are unit tests for the new code.
- [x] The new code is documented.
- [ ] The feature branch is rebased with respect to the target branch.

Steven-Roberts committed:
Improvements for IDA and KINSOL

Steven-Roberts committed:
Remove N_VGetArrayPointer_Serial calls

Steven-Roberts committed:
Resolve t_init_ issue

Steven-Roberts committed:
Add missing N_VDestroy

Steven-Roberts committed:
Apply pre-commmit fixes

pelesh commented on src/Solver/Dynamic/Ida.hpp:
```diff
@@ -95,96 +99,96 @@ namespace AnalysisManager
 
-      N_Vector yy_;  ///< Solution vector
-      N_Vector yp_;  ///< Solution derivatives vector
-      N_Vector tag_; ///< Tags differential variables
-      N_Vector q_;   ///< Integrand vector
```
I believe we need `tag_`. Otherwise code does not build:
```
/.../src/gridkit/GridKit/src/Solver/Dynamic/Ida.cpp: In member function 'int AnalysisManager::Sundials::Ida<ScalarT, IdxT>::initializeSimulation(real_type, bool)':
/.../src/gridkit/GridKit/src/Solver/Dynamic/Ida.cpp:189:13: error: 'tag_' was not declared in this scope
  189 |         if (tag_)
      |             ^~~~
make[2]: *** [src/Solver/Dynamic/CMakeFiles/solvers_dyn_obj_shared.dir/Ida.cpp.o] Error 1
```

pelesh commented on src/Solver/Dynamic/Ida.cpp:
```diff
@@ -43,18 +43,9 @@ namespace AnalysisManager
-      // IDAFree(&solver_);
+      SUNLinSolFree(linearSolver_);
+      SUNMatDestroy(JacobianMat_);
+      IDAFree(&solver_);
```
This looks like a bug in SUNDIALS. The backtrace of this segfault looks like:
```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x50)
  * frame #0: 0x0000000100123fbb libsundials_idas.6.dylib`IDAAdjFree(ida_mem=0x00007ff50b008200) at idaa.c:295:31
    frame #1: 0x0000000100115b5d libsundials_idas.6.dylib`IDAFree(ida_mem=0x00007ff50aa048d0) at idas.c:4064:3
    frame #2: 0x000000010009d3e4 libgridkit_solvers_dyn.0.dylib`AnalysisManager::Sundials::Ida<double, unsigned long>::~Ida(this=0x00007ff50aa048c0) at Ida.cpp:48:14
    frame #3: 0x000000010009d3f9 libgridkit_solvers_dyn.0.dylib`AnalysisManager::Sundials::Ida<double, unsigned long>::~Ida(this=0x00007ff50aa048c0) at Ida.cpp:50:5
    frame #4: 0x0000000100006698 dynconopt`main at DynamicConstrainedOpt.cpp:151:10
```
Going to `idaa.c`, the segfault seems to occur in this place:
```c++
   294      /* Delete check points one by one */
-> 295      while (IDAADJ_mem->ck_mem != NULL)
   296      {
   297        IDAAckpntDelete(&(IDAADJ_mem->ck_mem));
   298      }
```
It is hard to see anything suspicious on GridKit side.

Steven-Roberts replied:
Hmm, I was not able to replicate this. I'll see if I can spot a SUNDIALS bug, and if not, we may need to meet to walk through your code it in a debugger.

pelesh replied:
You need to enable Ipopt support in GridKit to build this example.

pelesh replied:
Issue fixed upstream.

pelesh commented on src/Solver/Dynamic/Ida.hpp:
```diff
@@ -26,6 +26,10 @@ namespace AnalysisManager
 
+      static_assert(std::is_same_v<real_type, sunrealtype>, "real_type must be the same type as sunrealtype");
+      // TODO: can't include this assert because cpp file instantiates a few values of IdxT
+      // static_assert(std::is_same_v<IdxT, sunindextype>, "IdxT must be the same type as sunindextype");
```
I think it is probably better to instantiate `Ida` only with data types available in SUNDIALS:
```c++
    template class Ida<sunrealtype, sunindextype>;
```
We don't need `static_assert` then.

pelesh commented on src/Solver/SteadyState/Kinsol.cpp:
```diff
@@ -173,40 +172,28 @@ namespace AnalysisManager
-      }
+      ScalarT* ydata = N_VGetArrayPointer(y);
+      std::copy(x.cbegin(), x.cend(), ydata);
     }
```
I don't think we need this for Kinsol at all.

pelesh commented on src/Solver/SteadyState/Kinsol.hpp:
```diff
@@ -125,17 +127,17 @@ namespace AnalysisManager
+      N_Vector yy_{};    ///< Solution vector
+      N_Vector scale_{}; ///< Scaling vector
+      N_Vector tag_{};   ///< Tags differential variables
+      N_Vector q_{};     ///< Integrand vector
```
I don't think there are differential variables to tag here.
```suggestion
      N_Vector scale_{}; ///< Scaling vector
      N_Vector q_{};     ///< Integrand vector
```

pelesh commented on src/Solver/Dynamic/Ida.cpp:
```diff
@@ -197,6 +189,7 @@ namespace AnalysisManager
         if (tag_)
           initType = IDA_YA_YDP_INIT;
 
+        // TODO: this doesn't doing anything without calling IDAGetConsistentIC
         retval = IDACalcIC(solver_, initType, 0.1);
         checkOutput(retval, "IDACalcIC");
```
I am not sure I understand this comment. My understanding is `IDACalcIC` overwrites solution vectors `yy_` and `yp_`.

Steven-Roberts replied:
I was thinking the consistent initial condition would need to be copied back into the model, but I realize now that's probably not necessary.

Steven-Roberts committed:
Fix some adjoint memory leaks

Steven-Roberts committed:
Remove unused function

Steven-Roberts committed:
merge

Steven-Roberts committed:
Merge branch 'sundials-updates' of github.com:ORNL/GridKit into sundials-updates

Steven-Roberts committed:
Update src/Solver/SteadyState/Kinsol.hpp

Co-authored-by: pelesh <peless@ornl.gov>

Steven-Roberts committed:
Use helper functions

pelesh commented:
@nkoukpaizan, we need to set SUNDIALS version in CI to `develop` branch > [07d21c2](https://github.com/LLNL/sundials/commit/07d21c2c5ae33211a9a2fafd8eac56de5582dce0).

nkoukpaizan committed:
sundials@develop in CI.

Steven-Roberts commented on .github/workflows/spack_default_build.yaml:
```diff
@@ -101,6 +101,7 @@ jobs:
         spack_spec:
           - gridkit@develop +enzyme
             ^enzyme@0.0.173
+            ^sundials@develop
```
This is fine for now, but as soon as the next SUNDIALS version is released, we should use that version. Breaking changes can occur on the develop branch.

pelesh commented:
The prior segfaults have been fixed with using SUNDIALS from `develop` branch. However, recent commits to the `sundials-updates` branch have introduced segfaults I have not seen before. The failing tests are all Ipopt related:
```
The following tests FAILED:
         14 - DynamicConOpt (SEGFAULT)
         15 - GenConstLoad (SEGFAULT)
         16 - GenInfiniteBus (SEGFAULT)
         17 - ParameterEst (SEGFAULT)
```
It seems there is an issue with getting data array pointer from the N_Vector. Full backtrace is available below:

<details>
Process 42256 launched: '/Users/55y/src/gridkit/build-gcc/examples/Experimental/DynamicConstrainedOpt/dynconopt' (x86_64)

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit https://github.com/coin-or/Ipopt
******************************************************************************

Process 42256 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x00000001000e46d4 libsundials_core.7.dylib`N_VGetArrayPointer(v=0x00007fe3fb909ae0) at sundials_nvector.c:338:13
   335 
   336  sunrealtype* N_VGetArrayPointer(N_Vector v)
   337  {
-> 338    if (v->ops->nvgetarraypointer)
   339    {
   340      return (sunrealtype*)v->ops->nvgetarraypointer(v);
   341    }
Target 0: (dynconopt) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
  * frame #0: 0x00000001000e46d4 libsundials_core.7.dylib`N_VGetArrayPointer(v=0x00007fe3fb909ae0) at sundials_nvector.c:338:13
    frame #1: 0x00000001000a285b libgridkit_solvers_dyn.0.dylib`AnalysisManager::Sundials::Ida<double, unsigned long>::initializeBackwardSimulation(double) [inlined] AnalysisManager::Sundials::Ida<double, unsigned long>::copyVec(x=size=2, y=0x00007fe3fb909ae0) at Ida.cpp:631:42
    frame #2: 0x00000001000a2856 libgridkit_solvers_dyn.0.dylib`AnalysisManager::Sundials::Ida<double, unsigned long>::initializeBackwardSimulation(this=0x00007fe3fb906110, tf=20) at Ida.cpp:378:14
    frame #3: 0x000000010004af24 libgridkit_solvers_opt.0.dylib`AnalysisManager::IpoptInterface::DynamicObjective<double, unsigned long>::eval_grad_f(this=0x00007fe3fb93da40, n=<unavailable>, x=0x00007fe3fb93fe00, new_x=<unavailable>, grad_f=0x00007fe3fb804fe0) at DynamicObjective.cpp:120:48
    frame #4: 0x00000001004fc94d libipopt.3.dylib`Ipopt::OrigIpoptNLP::grad_f(Ipopt::Vector const&) + 173
    frame #5: 0x00000001004c41c7 libipopt.3.dylib`Ipopt::IpoptCalculatedQuantities::curr_grad_f() + 359
    frame #6: 0x00000001004d143d libipopt.3.dylib`Ipopt::IpoptCalculatedQuantities::curr_grad_lag_x() + 685
    frame #7: 0x00000001004d2066 libipopt.3.dylib`Ipopt::IpoptCalculatedQuantities::curr_dual_infeasibility(Ipopt::ENormType) + 998
    frame #8: 0x00000001004f197e libipopt.3.dylib`Ipopt::OptimalityErrorConvergenceCheck::CheckConvergence(bool) + 78
    frame #9: 0x00000001004b9f60 libipopt.3.dylib`Ipopt::IpoptAlgorithm::Optimize(bool) + 272
    frame #10: 0x0000000100550592 libipopt.3.dylib`Ipopt::IpoptApplication::call_optimize() + 1858
    frame #11: 0x00000001005524be libipopt.3.dylib`Ipopt::IpoptApplication::OptimizeNLP(Ipopt::SmartPtr<Ipopt::NLP> const&, Ipopt::SmartPtr<Ipopt::AlgorithmBuilder>&) + 398
    frame #12: 0x000000010054e466 libipopt.3.dylib`Ipopt::IpoptApplication::OptimizeNLP(Ipopt::SmartPtr<Ipopt::NLP> const&) + 54
    frame #13: 0x0000000100006491 dynconopt`main at DynamicConstrainedOpt.cpp:96:34
(lldb) 
</details>

pelesh commented on src/Solver/Dynamic/Ida.cpp:
```diff
@@ -517,12 +506,17 @@ namespace AnalysisManager
     template <class ScalarT, typename IdxT>
     int Ida<ScalarT, IdxT>::deleteAdjoint()
     {
+      N_VDestroy(yyB_);
+      N_VDestroy(ypB_);
+      N_VDestroy(qB_);
+      SUNLinSolFree(linearSolverB_);
+      SUNMatDestroy(JacobianMatB_);
```
I guess I didn't document this well. Function `deleteAdjoint()` is only undoing adjoint initialization, so that adjoint can be initialized for a new number of backward integration points. Vectors `yyB_`, `ypB_`, etc., will be reused. Deleting them here causes a segfault at the second step of a dynamic constrained optimization. See for example this [function](https://github.com/ORNL/GridKit/blob/develop/src/Solver/Optimization/DynamicObjective.cpp#L103).
```suggestion
```
I am not sure how is this passing CI tests.

nkoukpaizan replied:
> I am not sure how is this passing CI tests.

CI does not build Ipopt because of the CoinHSL dependency. So, those 4 failing tests are not being tested in CI.

Steven-Roberts committed:
Remove todos

Steven-Roberts committed:
Fix segfault

pelesh commented on src/Solver/Dynamic/Ida.cpp:
```diff
@@ -39,22 +38,14 @@ namespace AnalysisManager
-      ///@todo this free is needed but on geninfbus this seg faults
-      // IDAFree(&solver_);
+      deleteQuadrature();
+      deleteAdjoint();
+      N_VDestroy(yyB_);
+      N_VDestroy(ypB_);
+      N_VDestroy(qB_);
+      SUNLinSolFree(linearSolverB_);
+      SUNMatDestroy(JacobianMatB_);
```
These will cause a segfault, if adjoint analysis was not configured and only forward simulation was run.

pelesh replied:
My bad, it seems SUNDIALS has a way of tracking what has been allocated and what has not, so these destroy functions appear to be safe.

pelesh reviewed (approved):
Good to merge. Consider merging #111 to `sundials-updates` before that.

pelesh committed:
Minor suggestions to SUNDIALS interface (#111)

* Minor suggestions to SUNDIALS interface

- Create "symmetric" functions for creating and deleting SUNDIALS objects.
- Modify RL example to test Ida destructor.


---------

Co-authored-by: pelesh <pelesh@users.noreply.github.com>
